homeassistant/components/mqtt/mixins.py

```diff
@@ -923,7 +938,7 @@ def _cleanup_discovery_on_remove(self) -> None:
             self._removed_from_hass = True
 
 
-def device_info_from_config(config) -> DeviceInfo | None:
+def device_info_from_config(config: DiscoveryInfoType | None) -> DeviceInfo | None:
```

Comment:
```suggestion
def device_info_from_config(config: dict[str, Any] | None) -> DeviceInfo | None:
```

Reply:
Maybe even rename the function:
```suggestion
def device_info_from_specifications(specifications: dict[str, Any] | None) -> DeviceInfo | None:
```